src/librustc/hir/lowering.rs

```diff
@@ -298,44 +294,103 @@ impl<'a> LoweringContext<'a> {
         }
     }
 
-    fn lower_path(&mut self,
-                  p: &Path,
-                  qself: Option<&hir::QSelf>,
-                  param_mode: ParamMode)
-                  -> hir::Path {
-        hir::Path {
+    fn lower_qpath(&mut self,
+                   id: NodeId,
+                   qself: &Option<QSelf>,
+                   p: &Path,
+                   param_mode: ParamMode)
+                   -> hir::QPath {
+        let qself_position = qself.as_ref().map(|q| q.position);
+        let qself = qself.as_ref().map(|q| self.lower_ty(&q.ty));
+
+        let resolution = self.resolver.get_resolution(id)
+                                      .unwrap_or(PathResolution::new(Def::Err));
+
+        let proj_start = p.segments.len() - resolution.depth;
+        let path = P(hir::Path {
             global: p.global,
-            segments: p.segments.iter().enumerate().map(|(i, segment)| {
-                let PathSegment { identifier, ref parameters } = *segment;
-                let param_mode = match (qself, param_mode) {
-                    (Some(qself), ParamMode::Optional) if i < qself.position => {
+            segments: p.segments[..proj_start].iter().enumerate().map(|(i, segment)| {
+                let param_mode = match (qself_position, param_mode) {
+                    (Some(j), ParamMode::Optional) if i < j => {
                         // This segment is part of the trait path in a
                         // qualified path - one of `a`, `b` or `Trait`
                         // in `<X as a::b::Trait>::T::U::method`.
                         ParamMode::Explicit
                     }
                     _ => param_mode
                 };
-                hir::PathSegment {
-                    name: identifier.name,
-                    parameters: self.lower_path_parameters(parameters, param_mode),
-                }
+                self.lower_path_segment(segment, param_mode)
+            }).collect(),
+            span: p.span,
+        });
+
+        // Simple case, either no projections, or only fully-qualified.
+        if resolution.depth == 0 {
```

Comment:
I completely forgot about it, the way things are set up it's not necessary. That is, `Trait::method` is really `Trait<..>::method::<..>` now and you could imagine `<..>` includes `Self.